src/kubernetes_cluster/proof/controller_runtime_safety.rs

```diff
@@ -46,6 +46,222 @@ pub open spec fn pending_req_has_unique_id<T>(cr_key: ObjectRef) -> StatePred<St
+    |s: State<T>| {
+        forall |msg: Message|
+            msg.content.is_APIRequest()
+            ==> #[trigger] s.message_in_flight_count(msg) <= 1
+    }
+}
+
+pub proof fn lemma_always_every_in_flight_req_is_unique<T>(reconciler: Reconciler<T>)
+    ensures
+        sm_spec(reconciler).entails(
+            always(lift_state(every_in_flight_req_is_unique::<T>()))
+        ),
+{
+    let invariant = every_in_flight_req_is_unique::<T>();
+    let stronger_next = |s, s_prime: State<T>| {
+        &&& next(reconciler)(s, s_prime)
+        &&& in_flight_req_has_lower_req_id()(s)
+    };
+    lemma_always_in_flight_req_has_lower_req_id::<T>(reconciler);
+    strengthen_next::<State<T>>(sm_spec(reconciler), next(reconciler), in_flight_req_has_lower_req_id(), stronger_next);
+    assert forall |s, s_prime: State<T>| invariant(s) && #[trigger] stronger_next(s, s_prime) implies invariant(s_prime) by {
+        assert forall |msg: Message| msg.content.is_APIRequest() implies #[trigger] s_prime.message_in_flight_count(msg) <= 1 by {
+            if (!s.message_in_flight(msg)) {
+                assert(s_prime.message_in_flight_count(msg) <= 1);
+            }
+            else {
+                assert(s.message_in_flight_count(msg) == 1);
+            }
+        };
+    };
+    init_invariant::<State<T>>(sm_spec(reconciler), init(reconciler), stronger_next, invariant);
+}
+
+pub open spec fn every_in_flight_msg_has_unique_id<T>() -> StatePred<State<T>> {
+    |s: State<T>| {
+        forall |msg: Message|
+            #[trigger] s.message_in_flight(msg)
+            ==> (
+                forall |other_msg: Message|
+                    #[trigger] s.message_in_flight(other_msg)
+                    && msg != other_msg
+                    ==> msg.content.get_msg_id() != other_msg.content.get_msg_id()
+            )
+    }
+}
+
+pub proof fn lemma_always_every_in_flight_msg_has_unique_id<T>(reconciler: Reconciler<T>)
+    ensures
+        sm_spec(reconciler).entails(
+            always(lift_state(every_in_flight_msg_has_unique_id::<T>()))
+        ),
+{
+    let invariant = every_in_flight_msg_has_unique_id::<T>();
+    let stronger_next = |s, s_prime: State<T>| {
+        next(reconciler)(s, s_prime)
+        && in_flight_req_has_lower_req_id::<T>()(s)
+        && in_flight_resp_has_lower_resp_id::<T>()(s)
+        && every_in_flight_req_is_unique::<T>()(s)
+    };
+    lemma_always_in_flight_req_has_lower_req_id::<T>(reconciler);
+    lemma_always_in_flight_resp_has_lower_resp_id::<T>(reconciler);
+    lemma_always_every_in_flight_req_is_unique::<T>(reconciler);
+    strengthen_next_3::<State<T>>(sm_spec(reconciler), next(reconciler), in_flight_req_has_lower_req_id::<T>(), in_flight_resp_has_lower_resp_id::<T>(), every_in_flight_req_is_unique::<T>(), stronger_next);
+    assert forall |s, s_prime: State<T>| invariant(s) && #[trigger] stronger_next(s, s_prime) implies invariant(s_prime) by {
```

Comment:
You can probably move all the proofs starting from this line to another proof function whose requires are inv(s) and stronger_next(s, s_prime) and whose ensures is inv(s_prime). This should reduce the verification time.